src/runtime/native/runtime.ml

```diff
@@ -101,24 +124,44 @@ let dump_counters_exn =
   Common.write_runtime_data
 
 let dump () =
-  match file_channel () with
-  (* Failure "TypeError: runtime.unix_open is not a function" indicates that the
-     code was compiled by js_of_ocaml. In this case, we *want* to do nothing at
-     process exit. See https://github.com/aantron/bisect_ppx/issues/334. *)
-  | exception Failure _ -> ()
-  | None -> ()
-  | Some channel ->
-      (try
-        dump_counters_exn channel
-      with _ ->
-        verbose Unable_to_write_file);
-      close_out_noerr channel
+  if not !bisect_file_written then
+    (if !sigterm_enable then
+       ignore @@ Unix.(sigprocmask SIG_BLOCK [Sys.sigterm]) ;
+     bisect_file_written := true;
+     match file_channel () with
+     (* Failure "TypeError: runtime.unix_open is not a function" indicates that the
+        code was compiled by js_of_ocaml. In this case, we *want* to do nothing at
+        process exit. See https://github.com/aantron/bisect_ppx/issues/334. *)
+     | exception Failure _ -> ()
+     | None -> ()
+     | Some channel ->
+       (try
+          dump_counters_exn channel
+        with _ ->
+          verbose Unable_to_write_file);
+       close_out_noerr channel)
+
+let sigterm_handler (_ : int) =
+  ignore @@ Unix.(sigprocmask SIG_BLOCK [Sys.sigterm]) ;
+  dump ();
+  exit 0
 
 let register_dump : unit Lazy.t =
   lazy (at_exit dump)
 
-let register_file ~bisect_file ~bisect_silent ~filename ~points =
+let register_file ~bisect_file ~bisect_silent ~bisect_sigterm ~filename ~points =
   (match bisect_file with None -> () | Some v -> default_bisect_file := v);
   (match bisect_silent with None -> () | Some v -> default_bisect_silent := v);
+  sigterm_enable := env_to_boolean "BISECT_SIGTERM" bisect_sigterm;
+  (if !sigterm_enable then
+    match Sys.(signal sigterm (Signal_handle sigterm_handler)) with
```

Comment:
this code is a bit ugly due to the warnings i've added, if they're not necessary then it can be simplified.